src/viz.py

```diff
@@ -0,0 +1,89 @@
+from typing import Literal
+import os
+import shutil
+from pathlib import Path
+import pandas as pd
+from matplotlib.pylab import plt
+from loguru import logger
+
+FIG_SIZE = (18, 9)
+
+
+class Visualization():
+    """Class for visualizing loss, evaluation metrics etc."""
+    
+    def __init__(self, output_dir:str):
+        # turn interactive plotting off
+        plt.ioff()
+
+        self.output_dir = Path(output_dir)
+
+        # create image folder
+        image_folder = self.output_dir / "images"
+
+        if not os.path.exists(image_folder):
+            os.makedirs(image_folder)
+        else:
+            shutil.rmtree(image_folder)
+            os.makedirs(image_folder)
+        
+        # self.df = pd.read_csv(train_history_path)
+        self.df = self._prepare_data()
+
+        self.column_mapping = {
+            "loss": ["train_loss", "eval_loss"],
+            "metrics": ["eval_wer", "eval_cer"]
+        }
+
+    def _prepare_data(self):
+        """Load the train_history.csv and clean up."""
+        df = pd.read_csv(self.output_dir / "train_history.csv")
+        
+        # get only columns we use for plot
+        use_columns = [
+            "loss", "learning_rate", "epoch", "step", "eval_loss", "eval_wer", 
+            "eval_cer"
+        ]
+        df = df[use_columns].rename(columns={'loss': 'train_loss'})
+
+        # handle duplicate rows for the same step by aggregating values
+        df = df.groupby('step').mean(numeric_only=True)
```

Comment:
this looks weird, but works. It's to handle the duplicates you see in the screenshot. 
it is just taking average of 1, so it is never changing the value while aligning all the mismatched rows.

<img width="789" alt="Screenshot 2024-12-06 at 2 21 30" src="https://github.com/user-attachments/assets/f2d311aa-2b1a-48dd-9c21-07d6ac404b31">